Cython/Utility/StringTools.c

```diff
@@ -395,9 +395,10 @@ static CYTHON_INLINE int __Pyx_GetItemInt_ByteArray_Fast(PyObject* string, Py_ss
         // The aim isn't to make the character read-writes atomic (although practically they probably are).
         // For simplicity, skip the critical section if we don't have the GIL. It's the user's problem!
         __Pyx_PyCriticalSection cs;
-        if (has_gil) { __Pyx_PyCriticalSection_Begin1(&cs, string); }
+        int lock = CYTHON_COMPILING_IN_CPYTHON_FREETHREADING && has_gil && (unsafe_shared || Py_REFCNT(o) > 1);
```

Comment:
This is specifically guarded by `CYTHON_COMPILING_IN_CPYTHON_FREETHREADING` because I'm worried that `Py_REFCNT` will be opaque in the Limited API so might not get skipped, even if the critical section block is obvious a no-op